### PR TITLE
chore: Simplify tox, bump dependencies

### PR DESCRIPTION
**Pull request by holmanb**
## Additional Context

- bump dependency versions (black, mypy, etc)
- fix mypy issues for newer versions
- since python3.7 is no longer supported by pytest, make a separate tox environment and CI job for the old version
- switch to Ubuntu 22.04 where needed for improved pip dependency resolution

## Merge type

- [ ] Squash merge using "Proposed Commit Message"
- [x] Rebase and merge unique commits. Requires commit messages per-commit each referencing the pull request number (#<PR_NUM>)

### CODE REVIEW

File: tox.ini
Comment (TheRealFalcon) [inline tox.ini:21]: Don't these next two sections kind of defeat the purpose of having multiple test environments? The reason there's multiple test environments is to minimize and isolate the dependencies needed to run the commands. If we don't see value in doing that (I don't see much...), then why not share a single virtualenv among all the commands rather than creating a separate one per command?

Even so, I would maybe want to keep the integration test requirements separate as those can be longer and fatter installs and not needed for most contributors.

One other thing to keep in mind is if we unconditionally install all test and integration requirements for every command we run, we do run the possibility of accidentally importing a test-only dependency in our code and then not catching it until sbuild time.

File: cloudinit/net/__init__.py
Comment (TheRealFalcon) [inline cloudinit/net/__init__.py:573]: This changes behavior, correct? Was the previous piece just wrong?

File: .github/workflows/unit.yml
Comment (TheRealFalcon) [inline .github/workflows/unit.yml:14]: We should continue to run a `lowest-supported` version with the set of lower-versioned dependencies to ensure that everything continues to work there.

**Review by TheRealFalcon — CHANGES_REQUESTED** [review]
I left some inline comments. I'm assuming that in the `black` commit, all of the changes there were auto-formatted? Is there anything there that you had to manually update?

File: pyproject.toml
Comment (holmanb) [reply pyproject.toml:279]: > I don't think we can move these here yet.
> Downstreams (including ours) generally run their unit tests against installed packages, not using tox. Focal's pytest doesn't support putting these options here.

Correct - on Focal pytest will continue to behave as it did before, and on >=Jammy pytest will start using these values. What's the problem with that? I can see an argument for dropping some of these values - the `testpaths` value is obviously unnecessary, but I don't think that any of these values are harmful, are they?

File: tox.ini
Comment (holmanb) [reply tox.ini:21]: > Don't these next two sections kind of defeat the purpose of having multiple test environments?

Yeah

> If we don't see value in doing that (I don't see much...)

Agreed.

> then why not share a single virtualenv among all the commands rather than creating a separate one per command?

I agree with you, however [it doesn't look like this is supported under tox 4](https://github.com/tox-dev/tox/issues/2788#issuecomment-1367437092). It wouldn't be difficult to wire up a test runner with a shared virtual environment with a more flexible test runner [like nox](https://github.com/wntrblm/nox), which allows creating a test runner which reuses virtual environments.

> Even so, I would maybe want to keep the integration test requirements separate as those can be longer and fatter installs and not needed for most contributors.

How much time / space do you think that these dependencies add to the initial run? After the environment is set up, it shouldn't add extra time since they are cached. I'm fine with that if you'd prefer that, but personally I'd lean towards something that's simpler to manage, since after the initial install for a particular environment it doesn't cost any extra time - and I'd bet that the different in install time / space isn't significant but maybe I'm wrong.

File: cloudinit/net/__init__.py
Comment (holmanb) [reply cloudinit/net/__init__.py:573]: > This changes behavior, correct? Was the previous piece just wrong?

No this behaves the same. It just defines `match` prior to initializing `cfg` - which allows mypy to reason about the type of `cfg` more correctly.

File: tox.ini
Comment (holmanb) [reply tox.ini:23]: > This gives us conflict requirements where we have pytest!=7.3.2 in test-requirements.txt, but pytest==7.0.1 here.

You are seeing a bug in older versions of pip - how did you hit this? pytest!=7.3.2 and pytest==7.0.1 are not mutually exclusive.

**Comment by holmanb** [comment]
> I left some inline comments. I'm assuming that in the `black` commit, all of the changes there were auto-formatted? 

Correct

> Is there anything there that you had to manually update?

[This commit](https://github.com/canonical/cloud-init/pull/5607/commits/db1d074b1888ca1ce846449c0ce22960240a359a) was manual changes by me for mypy, but otherwise no.

File: pyproject.toml
Comment (TheRealFalcon) [reply pyproject.toml:279]: Why will Focal pytest continue to behave as it did before? These used to be defined in tox.ini, which Focal's python will respect, but now they live in pyproject.toml, which Focal's pytest will not respect. I'm saying that I think the options need to continue to live in tox.ini until downstreams have caught up to allowing this configuration in pyproject.toml.

File: tox.ini
Comment (TheRealFalcon) [reply tox.ini:21]: > I agree with you, however https://github.com/tox-dev/tox/issues/2788#issuecomment-1367437092.

Sad

> How much time / space do you think that these dependencies add to the initial run? After the environment is set up, it shouldn't add extra time since they are cached. I'm fine with that if you'd prefer that, but personally I'd lean towards something that's simpler to manage, since after the initial install for a particular environment it doesn't cost any extra time - and I'd bet that the different in install time / space isn't significant but maybe I'm wrong.

It's not strongly held, but I also don't see one extra set of dependencies (that includes the other set) such that we have 2 basic configuration will make the overall tox file all that more complex. Caching is nice, but unless I'm using tox-uv (which is still buggy for me on 3.8), my env still winds up getting recreated more often than I would think is necessary.

File: cloudinit/net/__init__.py
Comment (TheRealFalcon) [reply cloudinit/net/__init__.py:573]: Ah sorry, I didn't see the line after the `if driver` had to changed too.

File: tox.ini
Comment (TheRealFalcon) [reply tox.ini:23]: > how did you hit this?

By running `tox` on Focal :smile: . It looks like this will install the latest version of pytest rather the one we have pinned.

File: pyproject.toml
Comment (holmanb) [reply pyproject.toml:279]: > Why will Focal pytest continue to behave as it did before?

Focal's pytest isn't invoked using tox, so running pytest on Focal didn't pick up these configuration settings before, and they won't pick up these configuration settings now. Nothing changes on Focal. And even if it did I don't think it would matter. These aren't even necessary, they are just reasonable defaults to have when debugging.

Comment (TheRealFalcon) [reply pyproject.toml:279]: > Focal's pytest isn't invoked using tox, so running pytest on Focal didn't pick up these configuration settings before, and they won't pick up these configuration settings now.

This isn't true. [Pytest uses tox.ini](https://docs.pytest.org/en/7.1.x/reference/customize.html#tox-ini) regardless of the existence of tox. It is a configuration file for pytest completely independent of tox.

>  These aren't even necessary, they are just reasonable defaults to have when debugging.

I don't think this is true. Without these settings we'll run integration tests by default when invoking pytest. Do we have something telling pybuild to not do that? We'll also get warnings about custom marks, but I suppose that isn't the end of the world.

File: tox.ini
Comment (holmanb) [reply tox.ini:23]: > > how did you hit this?
> 
> By running `tox` on Focal 😄 . It looks like this will install the latest version of pytest rather the one we have pinned.

I mistakenly assumed that we were actually using the pytest version somewhere but it looks like that wasn't in use before so I'll drop it and we can deal with that if/when pytest breaks.

File: pyproject.toml
Comment (holmanb) [reply pyproject.toml:279]: > > Focal's pytest isn't invoked using tox, so running pytest on Focal didn't pick up these configuration settings before, and they won't pick up these configuration settings now.
> 
> This isn't true. [Pytest uses tox.ini](https://docs.pytest.org/en/7.1.x/reference/customize.html#tox-ini) regardless of the existence of tox. It is a configuration file for pytest completely independent of tox.

Huh, weird

> > These aren't even necessary, they are just reasonable defaults to have when debugging.
> 
> I don't think this is true.

The lowest-supported ci test [passes with this configuration](https://github.com/canonical/cloud-init/actions/runs/10378310058/job/28734306492?pr=5607) in pytest:

```sh
+ tox -e lowest-supported
lowest-supported: install_deps> python -I -m pip install configobj==5.0.6 jinja2==2.10.1 jsonpatch==1.23 jsonschema==3.2.0 markupsafe==2.0.1 oauthlib==3.1.0 passlib pyserial==3.4 pytest-cov==2.8.1 pytest-mock==1.10.4 pytest==4.6.9 pyyaml==5.3.1 requests==2.22.0 responses==0.9.0 setuptools==45.2.0
lowest-supported: freeze> python -m pip freeze --all
lowest-supported: atomicwrites==1.4.1,attrs==24.2.0,certifi==2024.7.4,chardet==3.0.4,configobj==5.0.6,cookies==2.2.1,coverage==7.6.1,idna==2.8,Jinja2==2.10.1,jsonpatch==1.23,jsonpointer==3.0.0,jsonschema==3.2.0,MarkupSafe==2.0.1,more-itertools==10.4.0,oauthlib==3.1.0,packaging==24.1,passlib==1.7.4,pip==24.1,pluggy==0.13.1,py==1.11.0,pyrsistent==0.20.0,pyserial==3.4,pytest==4.6.9,pytest-cov==2.8.1,pytest-mock==1.10.4,PyYAML==5.3.1,requests==2.22.0,responses==0.9.0,setuptools==45.2.0,six==1.16.0,urllib3==1.25.11,wcwidth==0.2.13,wheel==0.43.0
lowest-supported: commands[0]> .tox/lowest-supported/bin/python -m pytest -m 'not hypothesis_slow' --cov=cloud-init --cov-branch tests/unittests
```

Note that the options defined in in `pyproject.toml` somehow got invoked, which is a bit mysterious, but presumably that's some sort of tox magic.

> Without these settings we'll run integration tests by default when invoking pytest. Do we have something telling pybuild to not do that? We'll also get warnings about custom marks, but I suppose that isn't the end of the world.

Our packaging invokes pytest using `make check`, which runs `python3 -m pytest -v tests/unittests cloudinit`.

So yeah, we'll get some warnings about custom marks, but that should be it. We're also used to ignoring deprecation warnings and so I'm leaning slightly towards making the change now since this is probably how we'll want it eventually anyways - but I can revert it if you'd prefer that.

File: tox.ini
Comment (holmanb) [reply tox.ini:21]: @TheRealFalcon I dropped the extra dependencies, let me know what you think.

File: pyproject.toml
Comment (TheRealFalcon) [reply pyproject.toml:279]: > I'm leaning slightly towards making the change now since this is probably how we'll want it eventually anyways - but I can revert it if you'd prefer that

I would prefer it. We don't **need** any of those pytest options, but they do add convenience. If we ever find we do need to update these options such that they are required on focal, that would require either migrating all of the options back to tox.ini or managing two configuration files. Additionally, I know of at least one downstream where they write and test code on the OS they're wanting to patch, and so I think it would nice to allow them to keep things as "old version compatible" as possible. 

If this required us to trade some kind of functionality I would have a different opinion, but AFAIK there's really no difference between putting pytest configuration in tox.ini vs pyproject.toml.

File: .github/workflows/cla.yml
Comment (TheRealFalcon) [inline .github/workflows/cla.yml:1]: Just curious, between this is and the labeler name, is there a reason you're preferring shorter names?

File: .github/workflows/unit.yml
Comment (TheRealFalcon) [inline .github/workflows/unit.yml:19]: What's this represent and why is it necessary? I tried Googling but failed.

File: .github/workflows/cla.yml
Comment (holmanb) [reply .github/workflows/cla.yml:1]: > Just curious, between this is and the labeler name, is there a reason you're preferring shorter names?

It's just a minor UI quality of life detail. In Github UI, these are truncated

image

image

Comment (holmanb) [reply .github/workflows/cla.yml:1]: > It's just a minor UI quality of life detail. In Github UI, these are truncated
> ...

I think that the following is slightly better as a result. With this change, it reduces truncation and increases signal to noise ratio by eliminating redundant data.

image

File: pyproject.toml
Comment (holmanb) [reply pyproject.toml:279]: > > I'm leaning slightly towards making the change now since this is probably how we'll want it eventually anyways - but I can revert it if you'd prefer that
> 
> I would prefer it.

Sounds good, +1

**Review by TheRealFalcon — APPROVED** [review]
LGTM assuming current checks pass

File: .github/workflows/cla.yml
Comment (TheRealFalcon) [reply .github/workflows/cla.yml:1]: Makes sense, thanks!

**Comment by holmanb** [comment]
> Thanks for all the changes! I'm +1 to the PR once the pytest configuration moves back to tox.ini

The review is much appreciated @TheRealFalcon! I just pushed a change which did that, and also reverted an unnecessary test addition to integration tests and fixed linkcheck which I had broken. I squashed changes before merging, so this is the `git diff origin/holmanb/simplify-tox` just before I force pushed.

```diff
diff --git a/pyproject.toml b/pyproject.toml
index 7eadd316a..2adba3761 100644
--- a/pyproject.toml
+++ b/pyproject.toml
@@ -275,27 +275,3 @@ lint.ignore = [
 
 [tool.ruff.lint.pydocstyle]
 convention = "pep257"
-
-[tool.pytest.ini_options]
-testpaths = "tools tests/unittests"
-addopts = "-vvvv --strict-markers --showlocals --durations 10"
-log_format = "%(asctime)s %(levelname)-9s %(name)s:%(filename)s:%(lineno)d %(message)s"
-log_date_format = "%Y-%m-%d %H:%M:%S"
-markers = [
-    "adhoc: only run on adhoc basis, not in any CI environment (travis or jenkins)",
-    "allow_all_subp: allow all subp usage (disable_subp_usage)",
-    "allow_subp_for: allow subp usage for the given commands (disable_subp_usage)",
-    "ci: run this integration test as part of CI test runs",
-    "ds_sys_cfg: a sys_cfg dict to be used by datasource fixtures",
-    "hypothesis_slow: hypothesis test too slow to run as unit test",
-    "instance_name: the name to be used for the test instance",
-    "integration_cloud_args: args for IntegrationCloud customization",
-    "is_iscsi: whether is an instance has iscsi net cfg or not",
-    "lxd_config_dict: set the config_dict passed on LXD instance creation",
-    "lxd_setup: specify callable to be called between init and start",
-    "lxd_use_exec: `execute` will use `lxc exec` instead of SSH",
-    "serial: tests that do not work in parallel, skipped with py3-fast",
-    "unstable: skip this test because it is flakey",
-    "user_data: the user data to be passed to the test instance",
-    "allow_dns_lookup: disable autochecking for host network configuration",
-]
diff --git a/tox.ini b/tox.ini
index bef415c4f..be5e1d647 100644
--- a/tox.ini
+++ b/tox.ini
@@ -197,7 +197,7 @@ commands = {envpython} -m sphinx -b spelling {posargs:-W doc/rtd doc/rtd_html}
 #
 # followed by manual verification of the links reported
 [testenv:linkcheck]
-deps = {[pinned_versions]deps}
+deps = -r{toxinidir}/doc-requirements.txt
 commands =
     {envpython} -m sphinx {posargs:-b linkcheck doc/rtd doc/rtd_html}
 
@@ -225,9 +225,7 @@ deps = {[latest_versions]deps}
 commands = {envpython} -m isort --check-only --diff {posargs:.}
 
 [testenv:integration-tests]
-deps =
-    {[pinned_versions]deps}
-    -r{toxinidir}/integration-requirements.txt
+deps = -r{toxinidir}/integration-requirements.txt
 commands = {envpython} -m pytest --log-cli-level=INFO {posargs:tests/integration_tests}
 passenv =
     CLOUD_INIT_*
@@ -236,9 +234,7 @@ passenv =
     OS_*
 
 [testenv:integration-tests-ci]
-deps =
-    {[pinned_versions]deps}
-    -r{toxinidir}/integration-requirements.txt
+deps = -r{toxinidir}/integration-requirements.txt
 commands = {envpython} -m pytest --log-cli-level=INFO {posargs:tests/integration_tests}
 passenv =
     CLOUD_INIT_*
@@ -250,9 +246,7 @@ setenv =
 [testenv:integration-tests-jenkins]
 # Pytest's RC=1 means "Tests were collected and run but some of the tests failed".
 # Do not fail in this case, but let Jenkins handle it using the junit report.
-deps =
-    {[pinned_versions]deps}
-    -r{toxinidir}/integration-requirements.txt
+deps = -r{toxinidir}/integration-requirements.txt
 allowlist_externals = sh
 commands = sh -c "{envpython} -m pytest --log-cli-level=INFO {posargs:tests/integration_tests/none} || [ $? -eq 1 ]"
 passenv =
@@ -266,6 +260,30 @@ passenv =
 setenv =
     PYTEST_ADDOPTS="-m not adhoc"
 
+[pytest]
+# TODO: s/--strict/--strict-markers/ once pytest version is high enough
+testpaths = tools tests/unittests
+addopts = --strict
+log_format = %(asctime)s %(levelname)-9s %(name)s:%(filename)s:%(lineno)d %(message)s
+log_date_format = %Y-%m-%d %H:%M:%S
+markers =
+    adhoc: only run on adhoc basis, not in any CI environment (travis or jenkins)
+    allow_all_subp: allow all subp usage (disable_subp_usage)
+    allow_subp_for: allow subp usage for the given commands (disable_subp_usage)
+    ci: run this integration test as part of CI test runs
+    ds_sys_cfg: a sys_cfg dict to be used by datasource fixtures
+    hypothesis_slow: hypothesis test too slow to run as unit test
+    instance_name: the name to be used for the test instance
+    integration_cloud_args: args for IntegrationCloud customization
+    is_iscsi: whether is an instance has iscsi net cfg or not
+    lxd_config_dict: set the config_dict passed on LXD instance creation
+    lxd_setup: specify callable to be called between init and start
+    lxd_use_exec: `execute` will use `lxc exec` instead of SSH
+    serial: tests that do not work in parallel, skipped with py3-fast
+    unstable: skip this test because it is flakey
+    user_data: the user data to be passed to the test instance
+    allow_dns_lookup: disable autochecking for host network configuration
+
 [coverage:paths]
 source =
     cloudinit/
```

**Pull request merged**
